feat: adds `samlidp gcp-workforce` command to teleport cli

flyinghermit opened the pull request:
A new command has been added to the teleport cli, which can be used to configure GCP workforce identity federation integration with Teleport SAML IdP.

`$ teleport integration configure samlidp gcp-workforce` The command is intended to set up a minimal workforce identity integration with our SAML IdP.

The command expects the following arguments:
- `--org-id`: GCP organization id
- `--pool-name`: name of the workforce pool
- `--pool-provider-name`: name of the workforce pool provider
- `--idp-metadata-url`: url of Teleport proxy that serves SAML IdP metadata. (<proxy-addr>/enterprise/saml-idp/metadata)

The API client does the following:
- Create workforce pool.
- Fetch Teleport SAML IdP metadata.
- Create workforce pool provider with Teleport SAML IdP metadata and attribute mapping.
  - Attribute mapping is added with just two attributes: `google.subject=assertion.subject` and `google.groups=assertion.attributes.roles`.


changelog: Added a new `teleport integration configure samlidp gcp-workforce` command to the teleport CLI which can be used to configure GCP Workforce Identity Federation integration with Teleport SAML IdP.

flyinghermit committed:
configure GCP workforce api client

flyinghermit committed:
add 'integration configure samlidp gcp-workforce' command to teleport ctl

flyinghermit committed:
update api to fetch metadata file from remote Teleport prox.
- minor comment copy and log copy edits

flyinghermit committed:
gcp resource name validation, checkandsetdefault funcs

flyinghermit committed:
use gcp-workforce subcommand path, rename gcpworkforce config as APIParams

flyinghermit committed:
return early on pending provider status

flyinghermit committed:
Merge branch 'master' into sshah/teleport-configure-gcp-wif

flyinghermit commented:
Manually tested in GCP cloud shell.

flyinghermit committed:
update comments and godoc

marcoandredinis commented on lib/integrations/samlidp/samlidpconfig/gcpworkforce.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	if ok := isValidGCPResourceName.MatchString(param); !ok {
+		return trace.BadParameter("resource name does not follow GCP resource naming convention.")
```
Returning the url, that you mentioned above (godoc), in the error message could be helpful.

flyinghermit replied:
That's a double edge sword. Link can change and we may spit out wrong link. I think the message "GCP resource naming convention" is verbose enough to search for naming convention.

marcoandredinis commented on tool/teleport/common/teleport.go:
```diff
@@ -1127,3 +1139,30 @@ func onIntegrationConfAccessGraphAWSSync(params config.IntegrationConfAccessGrap
+
+	gcpWorkforceService, err := samlidp.NewGCPWorkforceService(samlidp.GCPWorkforceService{
+		APIParams: params,
+		HTTPClient: &http.Client{
```
It seems we are creating a custom `CheckRedirect`. Is it really necessary?

Also, we have `defaults.HTTPClient()`. Can we use it here instead?

flyinghermit replied:
I don't want the client to follow redirect here and expect metadata to be available at the given endpoint. 

added comment to make this clear, also used `defaults.HTTPClient()` https://github.com/gravitational/teleport/pull/39820/commits/c65c9083aa11a02d3b89f1cb9e82181149cb3596

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,212 @@
+		Saml: &iam.GoogleIamAdminV1WorkforcePoolProviderSaml{
+			IdpMetadataXml: metadata,
+		},
+		AttributeMapping: map[string]string{
+			"google.subject": "assertion.subject",
+			"google.groups":  "assertion.attributes.roles",
+		},
```
Can we extract this and similar values to consts / package vars?

flyinghermit replied:
I am leaving this as is for now. I am thinking if we should expose attribute mapping as config param so definitely need to work on this in a followup.

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+// NewGCPWorkforceService creates a new GCPWorkforceService.
+func NewGCPWorkforceService(cfg GCPWorkforceService) (*GCPWorkforceService, error) {
```
Can we add test coverage for this method?

flyinghermit replied:
added in https://github.com/gravitational/teleport/pull/39820/commits/c65c9083aa11a02d3b89f1cb9e82181149cb3596

flyinghermit committed:
use slog

flyinghermit committed:
apply review suggestions:
- use 2*minute instead of 120*seconds
- use default http client and add redirect to the client
- NewGCPWorkforceService test

flyinghermit commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,212 @@
+// CreateWorkforcePoolAndProvider creates GCP Workforce Identity Federation pool and pool provider
+// with the given GCPWorkforceAPIParams values.
+func (s *GCPWorkforceService) CreateWorkforcePoolAndProvider(ctx context.Context) error {
+	fmt.Println("\nConfiguring Workforce Identity Federation pool and SAML provider.")
```
used slog https://github.com/gravitational/teleport/pull/39820/commits/bf8ea680bc48158b277ed8b4947c32bf17fdf55b

flyinghermit commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,212 @@
+// http.StatusForbidden or when context is canceled with timeout.
+func waitForPoolStatus(ctx context.Context, workforceService *iam.LocationsWorkforcePoolsService, poolName, poolDisplayName string) error {
+	fmt.Printf("Waiting for pool %q status to become available.\n", poolDisplayName)
+	ticker := time.NewTicker(1 * time.Second)
```
> But having a 1 second ticker would be too noisy

Yaa I started with printing at every ticker but ended up with this one. 

>What about changing this to a 5 or 10 second ticker?

This works but unless we make the subsequent log `still waiting...` or some progress message, I doubt it adds much value. The wait time is also around 1 min on average so I would not further delay the ticker here.

marcoandredinis replied:
Just having a log line with `still waiting...` every 10s would be better, IMO.
Having no output indicating progress for 1 minute might led people thinking that it crashed or something.

flyinghermit replied:
Fair enough on user expectation. In that case, I am updating the message "Waiting for pool status. It may take upto 2 minutes to become available." so user can expect that it takes some time.

flyinghermit replied:
https://github.com/gravitational/teleport/pull/39820/commits/fa31ce655094db82bdb255a5d806e075d7fbf49f

I have updated ticker to 5 seconds just to be less aggressive. The message says now "Waiting for pool status. It may take upto 2 minutes for new pool to become available." I think this is now verbose enough to set expectation without further complicating `waitForPoolStatus` func. 

Feel free to suggest if you think otherwise :)

flyinghermit commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,212 @@
+
+	"github.com/gravitational/trace"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
```
As of now, Workforce pool service is not available in the `v2` api - https://cloud.google.com/go/docs/reference/cloud.google.com/go/iam/latest/apiv2

I think v2 is still not fully ready yet and GCP only recommends to use it for using deny policies. 
Link https://cloud.google.com/iam/docs/reference/libraries#iam-v2
<img width="867" alt="Screenshot 2024-03-27 at 11 35 47 AM" src="https://github.com/gravitational/teleport/assets/10177602/56e786a2-ef20-4291-b2d1-e34dcf524aff">

r0mant replied:
@flyinghermit Could you add this comment on top of this import line? For future reference so we know what to check when it's time to migrate to v2.

flyinghermit committed:
update logger to use InfoContext. improve wait message, run fix-imports

flyinghermit committed:
update ticker to 5 second from 1 second

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+	}
+	workforceService := iam.NewLocationsWorkforcePoolsService(iamService)
+
+	slog.With("pool name", s.APIParams.PoolName).InfoContext(ctx, "Creating workforce pool.")
```
I think we should not use spaces in log fields
The rfd is not explicit about it, but the used examples seem to indicate so.
https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md#withfields

flyinghermit replied:
used `_` in spaces. looking at codebase its a mixed bag of cameCase and underscore but I think underscore is better for log readibility? https://github.com/gravitational/teleport/pull/39820/commits/d662a0b837c3a1ceb6224008ce391b2090668a72

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// NewGCPWorkforceService creates a new GCPWorkforceService.
+func NewGCPWorkforceService(cfg GCPWorkforceService) (*GCPWorkforceService, error) {
```
Receiving a type T and returning the same type T does not seem like the usual builder pattern we follow.
I would create a companion Config struct or just pass the `samlidpconfig.GCPWorkforceAPIParams` and the `http.Client`

flyinghermit replied:
I do not think there's anything wrong with receiving a type T and returning same type T. Builder pattern is also used for input validation which is what this func does.

FYI, I only added this new func so to add `checkAndSetdefault` for input validation. 

I have updated comment which now says "// NewGCPWorkforceService creates a new GCPWorkforceService with input validation." in https://github.com/gravitational/teleport/pull/39820/commits/91ae579c03203bf0d1d878c3d13b4ac51cb4d2ea if that helps.

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+
+// CreateWorkforcePoolAndProvider creates GCP Workforce Identity Federation pool and pool provider
+// with the given GCPWorkforceAPIParams values.
+func (s *GCPWorkforceService) CreateWorkforcePoolAndProvider(ctx context.Context) error {
```
Sorry, I actually meant to ask for tests in this method
It's the most complex part

flyinghermit replied:
FYI, sorry for not being clear in my PR description but I have intentionally avoided test in this PR.

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+// waitForPoolStatus waits for pool to come online. Returns immediately if error code is other than
+// http.StatusForbidden or when context is canceled with timeout.
+func waitForPoolStatus(ctx context.Context, workforceService *iam.LocationsWorkforcePoolsService, poolName, poolDisplayName string) error {
+	slog.InfoContext(ctx, "Waiting for pool status. It may take upto 2 minutes for new pool to become available.")
```
```suggestion
	slog.InfoContext(ctx, "Waiting for pool status. It may take up to 2 minutes for new pool to become available.")
```

marcoandredinis commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+						continue
+					}
+				}
+				return err
```
```suggestion
				return trace.Wrap(err)
```

flyinghermit committed:
use underscore in log field instead of space. add missing trace.Wrap in waitForPoolStatus

flyinghermit committed:
update NewGCPWorkforceService comment

r0mant commented on lib/integrations/samlidp/gcpworkforce.go:
```diff
@@ -0,0 +1,215 @@
+		// 2 minutes timeout is semi-random decision, chosen based on the fact that
+		// when creating workforce pool from the GCP web console, it mentions
+		// that the operation could take up to 2 minutes.
+		pollCtx, cancel := context.WithTimeout(ctx, 2*time.Minute)
```
I think 2 minutes is too short. I'd bump it to 5 minutes maybe.

Also, what happens if it fails to create within this time? What would a user do? Can they re-run the script?

r0mant replied:
What if the pool has actually been created but it took longer than 5 minutes? Will the command pick it up and proceed to create the provider?

flyinghermit replied:
I think the ideal solution here would be to have a `--pool` or `--pool-provider` flag so user can then manually create one or the other rather than having to delete and recreate. I will revisit this UX in a followup.

flyinghermit replied:
(Oh Github is not showing me new comments and then suddenly I have two new comments!)

> What if the pool has actually been created but it took longer than 5 minutes? Will the command pick it up and proceed to create the provider?

Right now - the command wont pick it up once its timed out, it exits. 

> Also, in your original bash script PR, we would just execute the "create pool" command and then proceed to creating provider right away. Does "gcloud" wait for the pool to be created before returning?

Yup, `gcloud` command exits only after pool is created and is available.

flyinghermit replied:
I have added 409 checker (Google api uses 409 response for conflicting resource name) if pool already exists so user have a slightly better ux in case the command timed out after creating pool, they can re run the command again and continue with pool provider. https://github.com/gravitational/teleport/pull/39820/commits/e4fad3c2858babca280a08fe103be6efc4af23c2

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -1127,3 +1139,34 @@ func onIntegrationConfAccessGraphAWSSync(params config.IntegrationConfAccessGrap
+	// Ensure we print output to the user. LogLevel at this point was set to Error.
+	utils.InitLogger(utils.LoggingForDaemon, slog.LevelInfo)
+
+	// httpClient is used to fetch SAML IdP metadata.
+	// we expect metadata to be available at the given SAMLIdPMetadataURL endpoint.
+	// As such client is configured not to follow redirect response.
+	httpClient, err := defaults.HTTPClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	httpClient.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+		return http.ErrUseLastResponse
+	}
```
Can this be a part of CheckAndSetDefaults on GCPWorkforceService?

flyinghermit committed:
review suggestions:
- increase wait timer to 5 mins
- iamv1 migration comment
- move http redirect checker to checkAndSet default

flyinghermit committed:
check for 409 response so that pool provider create op can continue if we timedout after creating pool

flyinghermit committed:
fix package import comment style to make linter happy

r0mant commented on tool/teleport/common/teleport.go:
```diff
@@ -1127,3 +1138,29 @@ func onIntegrationConfAccessGraphAWSSync(params config.IntegrationConfAccessGrap
+	// Ensure we print output to the user. LogLevel at this point was set to Error.
+	utils.InitLogger(utils.LoggingForDaemon, slog.LevelInfo)
+
+	// httpClient is used to fetch SAML IdP metadata.
+	httpClient, err := defaults.HTTPClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
I still think this should move to CheckAndSetDefaults cause it sets a default client.

flyinghermit replied:
https://github.com/gravitational/teleport/pull/39820/commits/07c52936198fcc729436fce7b3b6c5b8c81fe425

flyinghermit committed:
move default http client to gcpworkforce service checkandset defaults. update tests to relax on httpclient input param
